### PR TITLE
[16.0][MIG] mail parent recipient

### PR DESCRIPTION
**Pull request by imlopes**
Migrating mail_parent_recipient module from v15.0 to v16.0

### CODE REVIEW

**Review by mmequignon — APPROVED** [review]

**Review by flotho — APPROVED** [review]
code review,
LGTM

**Review by ivantodorovich — APPROVED** [review]
Nicpicking: You can merge the `pre-commit stuff` and `Migration to 16.0` commits, as there aren't any cosmetic changes. The only thing that's in pre-commit stuff is the version bump, which is actually the migration 😅

**Comment by i-vyshnevska** [comment]
These
Screenshot
 commits can be squashed

File: mail_parent_recipient/models/mail_template.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        results = super().generate_recipients(results, res_ids)
+        use_parent_address = (
+            self.env["ir.config_parameter"].sudo().get_param("mail.use_parent_address")
+        )
+        disabled = self.env.context.get("no_parent_mail_recipient")
+        if use_parent_address and not disabled:
+            for res_id, values in results.items():
+                partner_ids = values.get("partner_ids", [])
+                partners_with_emails = []
+                partners = self.env["res.partner"].sudo().browse(partner_ids)
```
Comment (hieulucky111) [inline mail_parent_recipient/models/mail_template.py:25]: can we put `res_partner_env = self.env["res.partner"]` at the top of function to improve a little bit performance.

File: mail_parent_recipient/models/mail_template.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                for partner in partners:
+                    if partner.email:
+                        partners_with_emails.append(partner.id)
+                    elif partner.commercial_partner_id.email:
```
Comment (hieulucky111) [inline mail_parent_recipient/models/mail_template.py:29]: we can combine like `elif partner.commercial_partner_id.email and partner.commercial_partner_id not in partners:`

File: mail_parent_recipient/tests/test_mail_parent_recipient.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+from odoo.addons.mail.tests.common import MailCommon
+
+
+class TestMailTemplate(MailCommon):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
+
+        cls.env["ir.config_parameter"].set_param("mail.use_parent_address", True)
```
Comment (hieulucky111) [inline mail_parent_recipient/tests/test_mail_parent_recipient.py:13]: should we use `sudo()` here

**Review by hieulucky111 — CHANGES_REQUESTED** [review]

**Comment by ajaniszewska-dev** [comment]
@imlopes would you find time to address latest comments so we merge this PR at some point in nearest future?

**Pull request closed without merging**